### PR TITLE
Workflow executions: Handle metadata and attachments

### LINKED ISSUE #495: WorkflowExecutions: Handling metadata and attachments
We are only allowing metadata to be flat, i.e. key value pairs with no arrays or hashes in the values.
* restrict the `metadata` field on samples_workflow_executions for this
* throw and handle/show errors gracefully

Metadata and attachments from workflows should be saved back to the samples
* metadata and output attachments exist on the samples_workflow_executions
  * the metadata should be merged with the sample metadata
  * the output attachments should be attached to the sample
    * do not duplicate these blobs

### PR DESCRIPTION
**Pull request by JeffreyThiessen**
## What does this PR do and why?
_Describe in detail what your merge request does and why._
Fixes #495 

Dependant on #501

During the CompletionService, add the following functionality
* attach outputs put onto SamplesWorkflowExecutions to the related Samples
  * these outputs use the same blob, so no file duplication
* merge output metadata to Sample objects
* Add step to flatten complex metadata arrays and hashes using dot notation

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Connect to a gh4gh_wes instance with related blob storage.
2. Run an workflow with some samples on it
3. Verify that after the completion step is finished the samples have attached output files
4. Verify that the attached output files are the same blob's as the blobs on the related SamplesWorkflowExecutions
5. Verify that the samples metadata has been updated

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

File: app/services/workflow_executions/completion_service.rb [hunk 1]
```diff
@@ -112,9 +117,32 @@ def process_samples_metadata(run_output_data:)
...
+        samples_workflow_execution.metadata = flatten(sample_metadata)
         samples_workflow_execution.save!
       end
     end
+
+    def merge_metadata_onto_samples
+      @workflow_execution.samples_workflow_executions&.each do |swe|
+        next if swe.metadata.nil?
+
+        params = { 'metadata' => swe.metadata }
```
Comment (ericenns) [inline app/services/workflow_executions/completion_service.rb:129]: **nitpick (blocking):** Can you add `analysis_id` to the params hash and set it to `@workflow_execution.id`. This way we can track that it was the `workflow_execution` that wrote this metadata.

File: test/services/workflow_executions/completion_service_test.rb
Comment (ericenns) [inline test/services/workflow_executions/completion_service_test.rb:1]: **nitpick (blocking):** Can you update the tests for writting Sample Metadata back to check the metadata_provenance and ensure it has the source of the workflow execution.

**Review by ericenns — CHANGES_REQUESTED** [review]
Just a few comments otherwise looking good.

File: test/helpers/metadata_helper_test.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+# require 'metadata_helper'
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply test/helpers/metadata_helper_test.rb]: fixed in https://github.com/phac-nml/irida-next/pull/504/commits/fecd6a50f494346e944bd756a5f2bfbd04c9579e

File: app/services/workflow_executions/completion_service.rb (hunk 1 above)
Comment (JeffreyThiessen) [reply app/services/workflow_executions/completion_service.rb]: added in https://github.com/phac-nml/irida-next/pull/504/commits/fecd6a50f494346e944bd756a5f2bfbd04c9579e

File: test/services/workflow_executions/completion_service_test.rb
Comment (JeffreyThiessen) [reply test/services/workflow_executions/completion_service_test.rb:1]: added in https://github.com/phac-nml/irida-next/pull/504/commits/fecd6a50f494346e944bd756a5f2bfbd04c9579e

**Review by ericenns — APPROVED** [review]
Looks good to me!

**Pull request merged**